tests/trainer/test_callbacks.py

```diff
@@ -0,0 +1,131 @@
+            self.on_init_start_called = False
+            self.on_init_end_called = False
+            self.on_epoch_start_called = False
+            self.on_epoch_end_called = False
+            self.on_batch_start_called = False
+            self.on_batch_end_called = False
+            self.on_train_start_called = False
+            self.on_train_end_called = False
+            self.on_validation_start_called = False
+            self.on_validation_end_called = False
+            self.on_test_start_called = False
+            self.on_test_end_called = False
+
+        def on_init_start(self, trainer):
+            assert isinstance(trainer, Trainer)
+            self.on_init_start_called = True
+
+        def on_init_end(self, trainer):
+            assert isinstance(trainer, Trainer)
+            self.on_init_end_called = True
+
+        def on_epoch_start(self, trainer, pl_module):
+            _check_args(trainer, pl_module)
+            self.on_epoch_start_called = True
+
+        def on_epoch_end(self, trainer, pl_module):
+            _check_args(trainer, pl_module)
+            self.on_epoch_end_called = True
+
+        def on_batch_start(self, trainer, pl_module):
+            _check_args(trainer, pl_module)
+            self.on_batch_start_called = True
+
+        def on_batch_end(self, trainer, pl_module):
+            _check_args(trainer, pl_module)
+            self.on_batch_end_called = True
+
+        def on_train_start(self, trainer, pl_module):
+            _check_args(trainer, pl_module)
+            self.on_train_start_called = True
+
+        def on_train_end(self, trainer, pl_module):
+            _check_args(trainer, pl_module)
+            self.on_train_end_called = True
+
+        def on_validation_start(self, trainer, pl_module):
+            _check_args(trainer, pl_module)
+            self.on_validation_start_called = True
+
+        def on_validation_end(self, trainer, pl_module):
+            _check_args(trainer, pl_module)
+            self.on_validation_end_called = True
+
+        def on_test_start(self, trainer, pl_module):
+            _check_args(trainer, pl_module)
+            self.on_test_start_called = True
+
+        def on_test_end(self, trainer, pl_module):
+            _check_args(trainer, pl_module)
+            self.on_test_end_called = True
+
+    test_callback = TestCallback()
+
+    trainer_options = {}
+    trainer_options['callbacks'] = [test_callback]
+    trainer_options['max_epochs'] = 1
+    trainer_options['val_percent_check'] = 0.1
+    trainer_options['train_percent_check'] = 0.2
+    trainer_options['show_progress_bar'] = False
+
+    assert not test_callback.on_init_start_called
+    assert not test_callback.on_init_end_called
+
+    # fit model
+    trainer = Trainer(**trainer_options)
+
+    assert trainer.callbacks[0] == test_callback
+    assert test_callback.on_init_start_called
+    assert test_callback.on_init_end_called
+    assert not test_callback.on_train_start_called
```

Comment:
shall all the other be tested if they are still False?
you may iterate over all atributes if you want with:
```
attribs = vars(test_callback)
for k in attribs:
    assert not attribs[k]
```

Reply:
i added more `assert`s but kept them explicit since it felt easier to read at a glance. what do you think?